File: JS/HashMap/SubstringConcatenation.js
module.exports = {
	//param A : string
	//param B : array of strings
	//return a array of integers
	findSubstring : function(A, B){
        var res = [];
        var map = {};
        B.forEach(function(str){
            if(!map.hasOwnProperty(str))map[str] = 0;
            map[str] = map[str] + 1;
        });

        var n = A.length, chunkSize = B[0].length;
        var len = B.length * chunkSize;
        for(var i = 0; i + len <= n; i++){
            var substr = A.substring(i, i + len); var allFound = true;
            for(var j = 0; j + chunkSize <= len; j += chunkSize){
                var chunkSub = substr.substring(j, j + chunkSize);
                if(!map.hasOwnProperty(chunkSub))allFound = false;
                else{
                    map[chunkSub] = map[chunkSub] - 1;
                    if(map[chunkSub] < 0)allFound = false;
                }
            }

            for(var j = 0; j + chunkSize <= len; j += chunkSize){
                var chunkSub = substr.substring(j, j + chunkSize);
                if(map.hasOwnProperty(chunkSub)){
                    map[chunkSub] = map[chunkSub] + 1;
                }
            }
            if(allFound)res.push(i);
        }
        //console.log(A.substring(98));
        return res;
	}
};

File: Java/HashMap/SubstringConcatenation.java
public class Solution {
	public ArrayList<Integer> findSubstring(String a, final List<String> b) {
	   ArrayList<Integer> res = new ArrayList<>();
	   HashMap<String, Integer> map = new HashMap<>();
	   for(String str : b){
	       if(!map.containsKey(str))map.put(str, 0);
	       map.put(str, map.get(str)+1);
	   }

	   // Now take substring of b's tot size
	   int chunkSize = b.get(0).length();
	   int len = b.size()*chunkSize;
	   for(int i = 0; i + len <= a.length(); i++){
	       String substr = a.substring(i, i + len);
	       boolean allFound = true;
	       //System.out.println(len+" "+substr);
	       for(int j = 0; j + chunkSize <= len; j += chunkSize){
	           //System.out.print("\t"+substr.substring(j, j + chunkSize));
	           String chunkSub = substr.substring(j, j + chunkSize);
	           if(!map.containsKey(chunkSub))allFound = false;
	           else{
	               map.put(chunkSub, map.get(chunkSub)-1);
	               int num = map.get(chunkSub);
	               if(num < 0)allFound = false;
	           }
	       }

	       // Add back removed num of occurence
	       for(int j = 0; j + chunkSize <= len; j += chunkSize){
	           String chunkSub = substr.substring(j, j + chunkSize);
	           if(map.containsKey(chunkSub)){
	               int num = map.get(chunkSub);
	               map.put(chunkSub, num+1);
	           }
	       }
	     if(allFound)res.add(i);
	   }
	   return res;
	}
}
